File: LeetCode/longest-common-subsequence.cpp
// https://leetcode.com/problems/longest-common-subsequence/
class Solution
{
public:
    int longestCommonSubsequence(string text1, string text2)
    {
        if (!text1.size() || !text2.size())
        {
            return 0;
        }
        vector<vector<int>> dp(text1.size() + 1, vector<int>(text2.size() + 1, 0));
        for (int i = 1; i < dp.size(); ++i)
        {
            for (int j = 1; j < dp[0].size(); ++j)
            {

                if (text1[i - 1] == text2[j - 1])
                {
                    // characters match, so remove it from both the strings
                    dp[i][j] = dp[i - 1][j - 1] + 1;
                }
                else
                {
                    // removing a character from one of the strings and seeing
                    // which removal gives a better result
                    dp[i][j] = max(dp[i][j - 1], dp[i - 1][j]);
                }
            }
        }
        return dp[dp.size() - 1][dp[0].size() - 1];
    }
};

File: LeetCode/longest-common-subsequence.java
// https://leetcode.com/problems/longest-common-subsequence/
// https://www.youtube.com/watch?v=ASoaQq66foQ

class Solution {
  public int longestCommonSubsequence(String text1, String text2) {
    if (text1.length() == 0 || text2.length() == 0) {
      return 0;
    }

    int[][] dp = new int[text1.length() + 1][text2.length() + 1];

    for (int i = 1; i < dp.length; ++i) {
      for (int j = 1; j < dp[0].length; ++j) {
        if (text1.charAt(i - 1) == text2.charAt(j - 1)) {
          dp[i][j] = 1 + dp[i - 1][j - 1];
        } else {
          dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
        }
      }
    }

    return dp[dp.length - 1][dp[0].length - 1];
  }
}